rlog/2025-07-05-zerokit-performance.md

```diff
@@ -0,0 +1,251 @@
+as is often the case with new technology, some problematic
+delays began to surface. Node recalculations, a common
+operation, were taking tens of seconds at the scales being
+tested and deployed—even exceeding 40 seconds at times.
+These delays accumulate, leading to total delays on the
+order of three hours under certain conditions.
+
+Naturally, we couldn't just let this sit. While we've
+touched on the issue of spam prevention, it's important to
+recognize that this technology is foundational that
+challenges conventional wisdom on how things must be done.
+Does the idea of "smart contracts without gas" catch your
+attention? Don't hold your breath just yet: the really
+interesting applications of this tech will be dead in the
+water, unless we can meet the challenge put to us.
+
+## The Challenge
+
+The plan of attack that the team put together was twofold:
+get rid of redundant operations and data taking up precious
+resources, and make the remaining operations go
+[*Blazingly Fast™*.](https://old.reddit.com/r/rust/comments/1avf1d8/blazingly_fast_memory_vulnerabilities_written_in/)
+
+Introducing the the star of the show for part 1: The main
+point of having this tree is to generate proofs so that
+peers can verify the claims being made. That doesn’t require
+the whole Merkle tree, just a single path, from leaf to
+root. The engineering work took us in a direction where
+these paths were the primary context in which ZK proofs
+operated, relegating the tree itself to an off-network
+reference. This reduced the burden imposed on the network
+significantly. Updating the data on the tree has similarly
+reduced, with the exception being that the siblings of each
+node were retained. This is called the **Stateless**
+approach.
+
+Well, stateless in the context of proof generation and
+verification. This is the critical context when it comes to
+performance, and the stateless approach does a great job,
+but these proofs have to come from *somewhere*. Each
+participant still needs to maintain the Merkle tree in their
+local environment. Without this tree, one cannot generate
+proofs or verify the proofs provided to them. Fortunately,
+one does not need to track the entire tree, but can be
+limited to a subset of the tree needed. With millions of
+participants on-chain, this can make the difference needed
+to make Zerokit empowered technologies accessible to those
+running raspberry Pis. Combine this with the fact that the
+heavy lifting operations of proof gen/verification being
+modular and separate, each participant can optimise to run
+things according to the strengths and requirements of their
+native hardware, easing the way to allow each participants
+to run their tree implementation at the speed of
+mach-ludicrous.
+
+Fortunately, the core of our already existing implementation
+was sane and well put together. Double-lucky for us, the
+talents of newly minted VAC/ACZ team member Sylvain were
```

Comment:
btw, Vinh also worked on this improvements. So if you wanna mention teammates by name, please add him too

or let's just refer to the whole team